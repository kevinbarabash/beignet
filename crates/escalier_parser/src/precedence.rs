use lazy_static::lazy_static;
use std::collections::hash_map::HashMap;

pub type Precedence = u32;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Associativity {
    Left,
    Right,
    Neither, // indicates a unary operator
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum Operator {
    // 18
    // Grouping, - parsed as an atom

    // 17
    MemberAccess,
    OptionalChaining,
    ComputedMemberAccess,
    NewWithArgumentList,
    FunctionCall,
    TemplateLiteral,

    // 16
    // NewWithoutArgList,

    // 15
    // PostfixIncrement,
    // PostfixDecrement,

    // 14
    LogicalNot,
    // BitwiseNot,
    UnaryPlus,
    UnaryMinus,
    // PrefixIncrement,
    // PrefixDecrement,
    Typeof,
    Void,
    Delete,
    Await,
    Throw,

    // 13
    Exponentiation,

    // 12
    Multiplication,
    Division,
    Remainder,

    // 11
    Addition,
    Subtraction,

    // 10
    // BitwiseLeftShift,
    // BitwiseRightShift,
    // BitwiseUnsignedRightShift,

    // 9
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    In,
    Instanceof,

    // 8
    Equals,    // always strict
    NotEquals, // always strict
    // StrictEquals,
    // StrictNotEquals,

    // 7
    // BitwiseAnd,

    // 6
    // BitwiseXor,

    // 5
    // BitwiseOr,

    // 4
    LogicalAnd,

    // 3
    LogicalOr,
    NullishCoalescing,

    // 2
    Assignment,
    Conditional,
    // Arrow, - superseded by `fn`
    Yield,
    YieldStar,
    Spread,
    // 1
    // CommaSequence,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OpInfo {
    prec: Precedence,
    assoc: Associativity,
}

impl OpInfo {
    pub fn new_prefix(prec: Precedence) -> Self {
        Self {
            prec,
            assoc: Associativity::Neither,
        }
    }

    pub fn new_infix(prec: Precedence, assoc: Associativity) -> Self {
        Self { prec, assoc }
    }

    pub fn new_postfix(prec: Precedence) -> Self {
        Self {
            prec,
            assoc: Associativity::Neither,
        }
    }

    pub fn normalized_prec(&self) -> Precedence {
        self.prec * 10
    }

    pub fn infix_postfix_prec(&self) -> Precedence {
        let normalized = self.normalized_prec();
        match self.assoc {
            Associativity::Left => normalized,
            Associativity::Right => normalized - 1,
            Associativity::Neither => normalized + 1,
        }
    }
}

lazy_static! {
    pub static ref PRECEDENCE_TABLE: HashMap<Operator, OpInfo> = {
        let mut table: HashMap<Operator, OpInfo> = HashMap::new();

        table.insert(
            Operator::MemberAccess,
            OpInfo::new_infix(17, Associativity::Left),
        );
        table.insert(
            Operator::OptionalChaining,
            OpInfo::new_infix(17, Associativity::Left),
        );
        table.insert(Operator::ComputedMemberAccess, OpInfo::new_postfix(17));
        table.insert(Operator::NewWithArgumentList, OpInfo::new_postfix(17));
        table.insert(Operator::FunctionCall, OpInfo::new_postfix(17));
        table.insert(Operator::TemplateLiteral, OpInfo::new_postfix(17));

        table.insert(Operator::LogicalNot, OpInfo::new_prefix(14));
        table.insert(Operator::UnaryPlus, OpInfo::new_prefix(14));
        table.insert(Operator::UnaryMinus, OpInfo::new_prefix(14));
        table.insert(Operator::Typeof, OpInfo::new_prefix(14));
        table.insert(Operator::Void, OpInfo::new_prefix(14));
        table.insert(Operator::Delete, OpInfo::new_prefix(14));
        table.insert(Operator::Await, OpInfo::new_prefix(14));
        table.insert(Operator::Throw, OpInfo::new_prefix(14));

        table.insert(
            Operator::Exponentiation,
            OpInfo::new_infix(13, Associativity::Right),
        );

        table.insert(
            Operator::Multiplication,
            OpInfo::new_infix(12, Associativity::Left),
        );
        table.insert(
            Operator::Division,
            OpInfo::new_infix(12, Associativity::Left),
        );
        table.insert(
            Operator::Remainder,
            OpInfo::new_infix(12, Associativity::Left),
        );

        table.insert(
            Operator::Addition,
            OpInfo::new_infix(11, Associativity::Left),
        );
        table.insert(
            Operator::Subtraction,
            OpInfo::new_infix(11, Associativity::Left),
        );

        table.insert(
            Operator::LessThan,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::LessThanOrEqual,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::GreaterThan,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(
            Operator::GreaterThanOrEqual,
            OpInfo::new_infix(9, Associativity::Left),
        );
        table.insert(Operator::In, OpInfo::new_infix(9, Associativity::Left));
        table.insert(
            Operator::Instanceof,
            OpInfo::new_infix(9, Associativity::Left),
        );

        table.insert(Operator::Equals, OpInfo::new_infix(8, Associativity::Left));
        table.insert(
            Operator::NotEquals,
            OpInfo::new_infix(8, Associativity::Left),
        );

        table.insert(
            Operator::LogicalAnd,
            OpInfo::new_infix(4, Associativity::Left),
        );

        table.insert(
            Operator::LogicalOr,
            OpInfo::new_infix(3, Associativity::Left),
        );
        table.insert(
            Operator::NullishCoalescing,
            OpInfo::new_infix(3, Associativity::Left),
        );

        table.insert(
            Operator::Assignment,
            OpInfo::new_infix(2, Associativity::Neither),
        );
        table.insert(
            Operator::Conditional,
            OpInfo::new_infix(2, Associativity::Left),
        );
        table.insert(Operator::Yield, OpInfo::new_infix(2, Associativity::Right));
        table.insert(
            Operator::YieldStar,
            OpInfo::new_infix(2, Associativity::Right),
        );
        table.insert(Operator::Spread, OpInfo::new_infix(2, Associativity::Right));

        table
    };
}
