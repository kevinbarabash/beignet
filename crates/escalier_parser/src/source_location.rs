use std::cmp::{max, min};
use std::ops::Range;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Position {
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct SourceLocation {
    pub start: Position,
    pub end: Position,
}

pub type Span = Range<usize>;

pub const DUMMY_SPAN: Span = Span { start: 0, end: 0 };

pub fn merge_spans(left: &Span, right: &Span) -> Span {
    Span {
        start: min(left.start, right.start),
        end: max(left.end, right.end),
    }
}
