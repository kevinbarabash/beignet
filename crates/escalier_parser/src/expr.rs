use crate::func_param::FuncParam;
use crate::literal::Literal;
use crate::pattern::Pattern;
use crate::source_location::SourceLocation;
use crate::stmt::Stmt;
use crate::type_ann::TypeAnn;

// TODO: track source location
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ObjectKey {
    Identifier(String), // TODO: use Identifier
    String(String),
    Number(String),
    Computed(Box<Expr>),
}

// TODO: track source location
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum Prop {
    Shorthand { key: String }, // TODO: use Identifier
    Property { key: ObjectKey, value: Expr },
    // TODO:
    // - method
    // - getter
    // - setter
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum PropOrSpread {
    Prop(Prop),
    Spread(Expr),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ExprOrSpread {
    Expr(Expr),
    Spread(Expr),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum ExprKind {
    Identifier(String),
    Literal(Literal),
    TemplateLiteral {
        parts: Vec<Literal>,
        exprs: Vec<Expr>,
    },
    Object {
        properties: Vec<PropOrSpread>,
    },
    Tuple {
        elements: Vec<ExprOrSpread>,
    },
    Binary {
        left: Box<Expr>,
        op: BinaryOp,
        right: Box<Expr>,
    },
    Unary {
        op: UnaryOp,
        right: Box<Expr>,
    },
    Index {
        left: Box<Expr>,
        right: Box<Expr>,
    },
    Function {
        params: Vec<FuncParam>,
        body: BlockOrExpr,
        type_ann: Option<TypeAnn>, // return type
    },
    Call {
        args: Vec<Expr>,
        callee: Box<Expr>,
    },
    Member {
        object: Box<Expr>,
        property: Box<Expr>,
    },
    OptionalChain {
        base: Box<Expr>,
    },
    IfElse {
        cond: Box<Expr>,
        consequent: Block,
        alternate: Option<Block>,
    },
    Match {
        expr: Box<Expr>,
        arms: Vec<MatchArm>,
    },
    Try {
        body: Block,
        // At least `catch` or `finally` must be present
        catch: Option<CatchClause>,
        finally: Option<Block>,
    },
    Do {
        body: Block,
    },
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct CatchClause {
    pub param: Option<Pattern>,
    pub body: Block,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct MatchArm {
    pub loc: SourceLocation,
    // pub span: Span,
    pub pattern: Pattern,
    pub guard: Option<Box<Expr>>,
    pub body: BlockOrExpr,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BlockOrExpr {
    Block(Block),
    Expr(Box<Expr>),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Block {
    pub loc: SourceLocation,
    pub stmts: Vec<Stmt>,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum BinaryOp {
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum UnaryOp {
    Plus,
    Minus,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Expr {
    pub kind: ExprKind,
    pub loc: SourceLocation,
}
