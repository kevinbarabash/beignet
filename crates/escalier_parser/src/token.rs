use crate::expr::Expr;
use crate::jsx::{JSXElement, JSXFragment};
use crate::source_location::SourceLocation;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TokenKind {
    Identifier(String), // [a-zA-Z_][a-zA-Z0-9_]*

    // Literals
    BoolLit(bool),
    NumLit(String),
    StrLit(String),
    StrTemplateLit {
        parts: Vec<Token>, // This should only contain StrLit tokens
        exprs: Vec<Expr>,
    },
    Null,
    Undefined,
    JSXElement(JSXElement),
    JSXFragment(JSXFragment),

    // Types
    Number,
    Boolean,
    String,
    Symbol,

    // Keywords
    Let,
    Return,
    Fn,
    Await,
    Async,
    If,
    Else,
    Mut, // denotes a binding to a mutable reference
    Var, // denotes a re-assignable binding
    Match,
    Try,
    Catch,
    Finally,
    Do,

    // Binary Operators
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equals,
    NotEquals,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Or,
    And,

    // Assignment Operators
    Assign,
    PlusAssign,
    MinusAssign,
    TimesAssign,
    DivideAssign,
    ModuloAssign,

    // punctuation
    Semicolon,
    Comma,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Arrow,
    Underscore,
    Question,
    QuestionDot, // used for optional chaining
    Dot,
    DotDot,    // used for ranges
    DotDotDot, // used for rest/spread
    Colon,

    Eof,
}

#[derive(Debug, PartialEq, Eq, Clone)]

pub struct Token {
    pub kind: TokenKind,
    pub loc: SourceLocation,
}
