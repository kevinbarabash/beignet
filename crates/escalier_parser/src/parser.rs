use std::iter::{Iterator, Peekable};

use crate::source_location::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Peekable<Box<dyn Iterator<Item = Token>>>,
}

impl Parser {
    pub fn new(tokens: Box<dyn Iterator<Item = Token>>) -> Self {
        Self {
            tokens: tokens.peekable(),
        }
    }

    pub fn next(&mut self) -> Token {
        self.tokens.next().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }

    pub fn peek(&mut self) -> Token {
        self.tokens.peek().cloned().unwrap_or(Token {
            kind: TokenKind::Eof,
            loc: SourceLocation {
                start: Position { line: 0, column: 0 },
                end: Position { line: 0, column: 0 },
            },
        })
    }
}
