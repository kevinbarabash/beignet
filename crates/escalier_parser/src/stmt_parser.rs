use crate::parser::*;
use crate::span::merge_spans;
use crate::stmt::{Stmt, StmtKind};
use crate::token::*;

impl<'a> Parser<'a> {
    pub fn parse_stmt(&mut self) -> Stmt {
        let token = self.peek().unwrap_or(&EOF).clone();

        match &token.kind {
            TokenKind::Let => {
                self.next().unwrap_or(EOF.clone());
                let pattern = self.parse_pattern();

                let type_ann = match self.peek().unwrap_or(&EOF).kind {
                    TokenKind::Colon => {
                        self.next().unwrap_or(EOF.clone());
                        Some(self.parse_type_ann())
                    }
                    _ => None,
                };

                assert_eq!(self.next().unwrap_or(EOF.clone()).kind, TokenKind::Assign);
                let expr = self.parse_expr();
                assert_eq!(
                    self.next().unwrap_or(EOF.clone()).kind,
                    TokenKind::Semicolon
                );

                let span = merge_spans(&token.span, &expr.get_span());
                Stmt {
                    kind: StmtKind::Let {
                        pattern,
                        expr,
                        type_ann,
                    },
                    span,
                }
            }
            TokenKind::Return => {
                self.next().unwrap_or(EOF.clone());
                let next = self.peek().unwrap_or(&EOF).clone();
                match next.kind {
                    TokenKind::Semicolon => {
                        self.next().unwrap_or(EOF.clone());
                        Stmt {
                            kind: StmtKind::Return { arg: None },
                            span: merge_spans(&token.span, &next.span),
                        }
                    }
                    _ => {
                        let arg = self.parse_expr();
                        assert_eq!(
                            self.next().unwrap_or(EOF.clone()).kind,
                            TokenKind::Semicolon
                        );

                        let span = merge_spans(&next.span, &arg.get_span());
                        Stmt {
                            kind: StmtKind::Return { arg: Some(arg) },
                            span,
                        }
                    }
                }
            }
            _ => {
                eprintln!("--- parse_stmt (expr) ---");
                let expr = self.parse_expr();
                assert_eq!(
                    self.next().unwrap_or(EOF.clone()).kind,
                    TokenKind::Semicolon
                );
                eprintln!("--- parse_stmt (assert semicolon) ---");

                let span = expr.get_span();
                Stmt {
                    kind: StmtKind::Expr { expr },
                    span,
                }
            }
        }
    }

    pub fn parse_program(&mut self) -> Vec<Stmt> {
        let mut stmts = Vec::new();
        while self.peek().unwrap_or(&EOF).kind != TokenKind::Eof {
            stmts.push(self.parse_stmt());
        }
        stmts
    }
}

pub fn parse(input: &str) -> Vec<Stmt> {
    let mut parser = Parser::new(input);
    parser.parse_program()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn single_statement() {
        let input = "let x = 5;";
        let stmts = parse(input);
        assert_eq!(stmts.len(), 1);
    }

    #[test]
    fn single_variable_expression() {
        let input = "x;";
        let stmts = parse(input);
        assert_eq!(stmts.len(), 1);
    }

    #[test]
    fn multiple_statements() {
        let input = r#"
        let x = 5;
        let y = 10;
        x + y;
        return; 
        "#;

        let stmts = parse(input);

        assert_eq!(stmts.len(), 4);
    }

    #[test]
    fn parse_let() {
        insta::assert_debug_snapshot!(parse(r#"let y = m*x + b;"#));
    }

    #[test]
    fn parse_let_with_type_annotation() {
        insta::assert_debug_snapshot!(parse(r#"let y: number = m*x + b;"#));
    }

    #[test]
    fn parse_let_with_destructuring() {
        insta::assert_debug_snapshot!(parse(r#"let {x, y} = point;"#));
    }

    #[test]
    fn parse_let_with_destructuring_and_type_annotation() {
        insta::assert_debug_snapshot!(parse(r#"let {x, y}: Point = point;"#));
    }

    #[test]
    fn parse_assignment() {
        insta::assert_debug_snapshot!(parse(r#"y = m*x + b;"#));
        insta::assert_debug_snapshot!(parse(r#"p.x = 5;"#));
        insta::assert_debug_snapshot!(parse(r#"p["y"] = 10;"#));
    }

    #[test]
    fn parse_conditionals() {
        insta::assert_debug_snapshot!(parse("let max = if (x > y) { x; } else { y; };"));
        insta::assert_debug_snapshot!(parse("if (foo) { console.log(foo); };"));
    }

    #[test]
    fn parse_lambda() {
        insta::assert_debug_snapshot!(parse("let add = fn (x, y) => x + y;"));
        insta::assert_debug_snapshot!(parse("let add = fn (x) => fn (y) => x + y;"));
    }

    #[test]
    fn parse_let_destructuring() {
        insta::assert_debug_snapshot!(parse("let {x, y} = point;"));
        insta::assert_debug_snapshot!(parse("let {x: x1, y: y1} = p1;"));
        insta::assert_debug_snapshot!(parse("let [p1, p2] = line;"));
        insta::assert_debug_snapshot!(parse("let [head, ...tail] = polygon;"));
    }

    #[test]
    fn parse_let_fn_with_fn_type() {
        insta::assert_debug_snapshot!(parse(
            r#"let add: fn (a: number, b: number) => number = fn (a, b) => a + b;"#
        ));
    }
}
