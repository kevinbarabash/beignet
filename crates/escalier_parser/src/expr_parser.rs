use crate::expr::{BinaryOp, Expr, ExprKind, UnaryOp};
use crate::parser::Parser;
use crate::precedence::{Associativity, Operator, PRECEDENCE_TABLE};
use crate::source_location::*;
use crate::stmt::Stmt;
use crate::stmt_parser::parse_stmt;
use crate::token::{Token, TokenKind};

fn get_prefix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        TokenKind::Plus => PRECEDENCE_TABLE.get(&Operator::UnaryPlus).cloned(),
        TokenKind::Minus => PRECEDENCE_TABLE.get(&Operator::UnaryMinus).cloned(),
        _ => None,
    }
}

fn get_infix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        // multiplicative
        TokenKind::Times => PRECEDENCE_TABLE.get(&Operator::Multiplication).cloned(),
        TokenKind::Divide => PRECEDENCE_TABLE.get(&Operator::Division).cloned(),
        // TODO: modulo

        // additive
        TokenKind::Plus => PRECEDENCE_TABLE.get(&Operator::Addition).cloned(),
        TokenKind::Minus => PRECEDENCE_TABLE.get(&Operator::Subtraction).cloned(),

        // equality
        TokenKind::Equals => PRECEDENCE_TABLE.get(&Operator::Equals).cloned(),
        TokenKind::LessThan => PRECEDENCE_TABLE.get(&Operator::LessThan).cloned(),
        TokenKind::LessThanOrEqual => PRECEDENCE_TABLE.get(&Operator::LessThanOrEqual).cloned(),
        TokenKind::GreaterThan => PRECEDENCE_TABLE.get(&Operator::GreaterThan).cloned(),
        TokenKind::GreaterThanOrEqual => {
            PRECEDENCE_TABLE.get(&Operator::GreaterThanOrEqual).cloned()
        }

        // logic
        TokenKind::And => PRECEDENCE_TABLE.get(&Operator::LogicalAnd).cloned(),
        TokenKind::Or => PRECEDENCE_TABLE.get(&Operator::LogicalOr).cloned(),

        // assignment
        TokenKind::Assign => PRECEDENCE_TABLE.get(&Operator::Assignment).cloned(),
        _ => None,
    }
}

fn get_postfix_precedence(op: &Token) -> Option<(u8, Associativity)> {
    match &op.kind {
        TokenKind::LeftBracket => PRECEDENCE_TABLE
            .get(&Operator::ComputedMemberAccess)
            .cloned(),
        TokenKind::LeftParen => PRECEDENCE_TABLE.get(&Operator::FunctionCall).cloned(),
        // TODO: handle optional chaining
        TokenKind::Dot => PRECEDENCE_TABLE.get(&Operator::MemberAccess).cloned(),
        _ => None,
    }
}

fn parse_params(parser: &mut Parser) -> Vec<String> {
    let mut params = Vec::new();
    while parser.peek().kind != TokenKind::RightParen {
        let param = parser.next();
        if let TokenKind::Identifier(name) = param.kind {
            params.push(name.to_owned());
        } else {
            panic!("Expected identifier, got {:?}", param);
        }

        match parser.peek().kind {
            TokenKind::RightParen => break,
            TokenKind::Comma => {
                parser.next();
            }
            _ => panic!("Expected comma or right paren, got {:?}", parser.peek()),
        }
    }
    params
}

fn parse_block(parser: &mut Parser) -> Vec<Stmt> {
    let mut stmts = Vec::new();
    while parser.peek().kind != TokenKind::RightBrace {
        stmts.push(parse_stmt(parser));
    }
    stmts
}

fn parse_expr_with_precedence(parser: &mut Parser, precedence: u8) -> Expr {
    let next = parser.next();

    let mut lhs = match &next.kind {
        TokenKind::Number(n) => Expr {
            kind: ExprKind::Number(n.to_owned()),
            loc: next.loc.clone(),
        },
        TokenKind::Identifier(id) => Expr {
            kind: ExprKind::Identifier(id.to_owned()),
            loc: next.loc.clone(),
        },
        TokenKind::LeftParen => {
            let lhs = parse_expr_with_precedence(parser, 0);
            assert_eq!(parser.next().kind, TokenKind::RightParen);
            lhs
        }
        TokenKind::Fn => {
            assert_eq!(parser.next().kind, TokenKind::LeftParen);
            let params = parse_params(parser);
            assert_eq!(parser.next().kind, TokenKind::RightParen);
            assert_eq!(parser.next().kind, TokenKind::LeftBrace);
            let body = parse_block(parser);
            let close_brace = parser.next();
            assert_eq!(close_brace.kind, TokenKind::RightBrace);

            let loc = merge_locations(&next.loc, &close_brace.loc);
            Expr {
                kind: ExprKind::Function { params, body },
                loc,
            }
        }
        t => match get_prefix_precedence(&next) {
            Some(precendence) => {
                let op = match t {
                    TokenKind::Plus => UnaryOp::Plus,
                    TokenKind::Minus => UnaryOp::Minus,
                    _ => panic!("unexpected token: {:?}", t),
                };

                let rhs = parse_expr_with_precedence(parser, precendence.0);
                let loc = merge_locations(&next.loc, &rhs.loc);
                Expr {
                    kind: ExprKind::Unary {
                        op,
                        right: Box::new(rhs),
                    },
                    loc,
                }
            }
            None => panic!("unexpected token: {:?}", t),
        },
    };

    loop {
        let next = parser.peek();
        if let TokenKind::Eof = next.kind {
            break;
        }

        if let Some(next_precedence) = get_postfix_precedence(&next) {
            if precedence >= next_precedence.0 {
                break;
            }

            let precedence = if next_precedence.1 == Associativity::Left {
                next_precedence.0
            } else {
                next_precedence.0 - 1
            };

            parser.next();

            lhs = match &next.kind {
                TokenKind::LeftBracket => {
                    let rhs = parse_expr_with_precedence(parser, 0);
                    let loc = merge_locations(&lhs.loc, &rhs.loc);
                    assert_eq!(parser.next().kind, TokenKind::RightBracket);
                    Expr {
                        kind: ExprKind::Index {
                            left: Box::new(lhs),
                            right: Box::new(rhs),
                        },
                        loc,
                    }
                }
                TokenKind::LeftParen => {
                    let mut args = Vec::new();
                    while parser.peek().kind != TokenKind::RightParen {
                        args.push(parse_expr_with_precedence(parser, 0));

                        match parser.peek().kind {
                            TokenKind::RightParen => break,
                            TokenKind::Comma => {
                                parser.next();
                            }
                            _ => panic!("Expected comma or right paren, got {:?}", parser.peek()),
                        }
                    }
                    let loc = merge_locations(&lhs.loc, &parser.peek().loc);
                    assert_eq!(parser.next().kind, TokenKind::RightParen);
                    Expr {
                        kind: ExprKind::Call {
                            callee: Box::new(lhs),
                            args,
                        },
                        loc,
                    }
                }
                TokenKind::Dot => {
                    let rhs = parse_expr_with_precedence(parser, precedence);
                    let loc = merge_locations(&lhs.loc, &rhs.loc);
                    Expr {
                        kind: ExprKind::Member {
                            object: Box::new(lhs),
                            property: Box::new(rhs),
                        },
                        loc,
                    }
                }
                _ => panic!("unexpected token: {:?}", next),
            };

            continue;
        }

        if let Some(next_precedence) = get_infix_precedence(&next) {
            if precedence >= next_precedence.0 {
                break;
            }

            parser.next();

            let op: BinaryOp = match &next.kind {
                TokenKind::Plus => BinaryOp::Plus,
                TokenKind::Minus => BinaryOp::Minus,
                TokenKind::Times => BinaryOp::Times,
                TokenKind::Divide => BinaryOp::Divide,
                TokenKind::Equals => BinaryOp::Equals,
                TokenKind::NotEquals => BinaryOp::NotEquals,
                TokenKind::LessThan => BinaryOp::LessThan,
                TokenKind::LessThanOrEqual => BinaryOp::LessThanOrEqual,
                TokenKind::GreaterThan => BinaryOp::GreaterThan,
                TokenKind::GreaterThanOrEqual => BinaryOp::GreaterThanOrEqual,
                TokenKind::And => BinaryOp::And,
                TokenKind::Or => BinaryOp::Or,
                _ => panic!("unexpected token: {:?}", next),
            };

            let precedence = if next_precedence.1 == Associativity::Left {
                next_precedence.0
            } else {
                next_precedence.0 - 1
            };

            let rhs = parse_expr_with_precedence(parser, precedence);
            let loc = merge_locations(&lhs.loc, &rhs.loc);

            lhs = Expr {
                kind: ExprKind::Binary {
                    op,
                    left: Box::new(lhs),
                    right: Box::new(rhs),
                },
                loc,
            };

            continue;
        }

        break;
    }

    lhs
}

pub fn parse_expr(parser: &mut Parser) -> Expr {
    parse_expr_with_precedence(parser, 0)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lexer::Lexer;

    pub fn parse(input: &str) -> Expr {
        let mut lexer = Lexer::new(input);
        let tokens = lexer.lex();
        let mut parser = Parser::new(tokens);
        parse_expr(&mut parser)
    }

    #[test]
    fn parse_simple_addition() {
        insta::assert_debug_snapshot!(parse("1 + 2 + 3"));
    }

    #[test]
    fn parse_addition_and_subtraction() {
        insta::assert_debug_snapshot!(parse("1 + 2 - 3 + 4"));
    }

    #[test]
    fn parse_additive_and_multiplicative() {
        insta::assert_debug_snapshot!(parse("1 * 2 + 3"));
    }

    #[test]
    fn parse_parens() {
        insta::assert_debug_snapshot!(parse("5 * (x + 1)"));
    }

    #[test]
    fn parse_comparisons_and_logic() {
        insta::assert_debug_snapshot!(parse("a > b && c >= d || e < f && g <= h"));
    }

    #[test]
    fn parse_unary_operators() {
        insta::assert_debug_snapshot!(parse("--a - +b"));
    }

    #[test]
    fn parse_indexing() {
        insta::assert_debug_snapshot!(parse("a[1][c]"));
    }

    #[test]
    fn parse_function() {
        let src = r#"fn () { let x = 5; let y = 10; return x + y; }"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn parse_function_with_params() {
        let src = r#"fn (x, y) { return x + y; }"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    #[should_panic]
    fn parse_function_expected_comma_or_left_brace() {
        let src = r#"fn (x, y { return x + y; }"#;
        parse(src);
    }

    #[test]
    #[should_panic]
    fn parse_function_expected_identifier() {
        let src = r#"fn (, y) { return x + y; }"#;
        parse(src);
    }

    #[test]
    fn parse_function_call() {
        let src = r#"add(5 * i, 10 * j)"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn parse_call_expr() {
        let src = r#"foo[bar](5, 10)"#;
        insta::assert_debug_snapshot!(parse(src));
    }

    #[test]
    fn parse_member_access() {
        insta::assert_debug_snapshot!(parse("a.b.c"));
        insta::assert_debug_snapshot!(parse("a.b+c.d"));
    }
}
