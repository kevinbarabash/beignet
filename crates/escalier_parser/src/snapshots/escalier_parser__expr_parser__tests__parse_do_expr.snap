---
source: crates/escalier_parser/src/expr_parser.rs
expression: "parse(r#\"\n            do {\n                let x = 5\n                let y = 10\n                x + y\n            }\n            \"#)"
---
Expr {
    kind: Do(
        Do {
            body: Block {
                span: 15..106,
                stmts: [
                    Stmt {
                        kind: Let {
                            is_declare: false,
                            is_var: false,
                            pattern: Pattern {
                                kind: Ident(
                                    BindingIdent {
                                        name: "x",
                                        span: 38..39,
                                        mutable: false,
                                    },
                                ),
                                span: 38..39,
                                inferred_type: None,
                            },
                            expr: Some(
                                Expr {
                                    kind: Num(
                                        Num {
                                            value: "5",
                                        },
                                    ),
                                    span: 42..43,
                                    inferred_type: None,
                                },
                            ),
                            type_ann: None,
                        },
                        span: 34..43,
                        inferred_type: None,
                    },
                    Stmt {
                        kind: Let {
                            is_declare: false,
                            is_var: false,
                            pattern: Pattern {
                                kind: Ident(
                                    BindingIdent {
                                        name: "y",
                                        span: 64..65,
                                        mutable: false,
                                    },
                                ),
                                span: 64..65,
                                inferred_type: None,
                            },
                            expr: Some(
                                Expr {
                                    kind: Num(
                                        Num {
                                            value: "10",
                                        },
                                    ),
                                    span: 68..70,
                                    inferred_type: None,
                                },
                            ),
                            type_ann: None,
                        },
                        span: 60..70,
                        inferred_type: None,
                    },
                    Stmt {
                        kind: Expr {
                            expr: Expr {
                                kind: Binary(
                                    Binary {
                                        left: Expr {
                                            kind: Ident(
                                                Ident {
                                                    name: "x",
                                                    span: 87..88,
                                                },
                                            ),
                                            span: 87..88,
                                            inferred_type: None,
                                        },
                                        op: Plus,
                                        right: Expr {
                                            kind: Ident(
                                                Ident {
                                                    name: "y",
                                                    span: 91..92,
                                                },
                                            ),
                                            span: 91..92,
                                            inferred_type: None,
                                        },
                                    },
                                ),
                                span: 87..92,
                                inferred_type: None,
                            },
                        },
                        span: 87..92,
                        inferred_type: None,
                    },
                ],
            },
        },
    ),
    span: 13..106,
    inferred_type: None,
}
